java/ql/src/semmle/code/java/frameworks/spring/SpringWebUtil.qll

```diff
@@ -0,0 +1,241 @@
+        "org.springframework.web.util;UriBuilder;true;replaceQueryParam;(String,Collection);;Element of Argument[1];Argument[-1];taint",
+        "org.springframework.web.util;UriBuilder;true;replaceQueryParam;(String,Object[]);;ArrayElement of Argument[1];ReturnValue;taint",
+        "org.springframework.web.util;UriBuilder;true;replaceQueryParam;(String,Object[]);;ArrayElement of Argument[1];Argument[-1];taint",
+        "org.springframework.web.util;UriBuilder;true;replaceQueryParams;;;Argument[-1];ReturnValue;value",
+        "org.springframework.web.util;UriBuilder;true;replaceQueryParams;;;Element of MapKey of Argument[0];ReturnValue;taint",
+        "org.springframework.web.util;UriBuilder;true;replaceQueryParams;;;Element of MapValue of Argument[0];ReturnValue;taint",
+        "org.springframework.web.util;UriBuilder;true;scheme;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;UriBuilder;true;scheme;;;Argument[0];ReturnValue;taint",
+        "org.springframework.web.util;UriBuilder;true;scheme;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;UriBuilder;true;userInfo;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;UriBuilder;true;userInfo;;;Argument[0];ReturnValue;taint",
+        "org.springframework.web.util;UriBuilder;true;userInfo;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;UriBuilderFactory;true;builder;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;UriBuilderFactory;true;uriString;;;Argument[-1..0];ReturnValue;taint",
+        "org.springframework.web.util;UriComponents$UriTemplateVariables;true;getValue;;;MapValue of Argument[-1];ReturnValue;value",
+        "org.springframework.web.util;UriTemplateHandler;true;expand;;;Argument[-1..0];ReturnValue;taint",
+        "org.springframework.web.util;UriTemplateHandler;true;expand;(String,Map);;MapValue of Argument[1];ReturnValue;taint",
+        "org.springframework.web.util;UriTemplateHandler;true;expand;(String,Object[]);;ArrayElement of Argument[1];ReturnValue;taint",
+        "org.springframework.web.util;AbstractUriTemplateHandler;true;getBaseUrl;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;AbstractUriTemplateHandler;true;setBaseUrl;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;AbstractUriTemplateHandler;true;setDefaultUriVariables;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;ContentCachingRequestWrapper;false;ContentCachingRequestWrapper;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;ContentCachingRequestWrapper;false;getContentAsByteArray;;;Argument[-1];ReturnValue;taint",
+        // note for review: the `content` of this wrapper is determined by what's written to it; there's a function `getWriter()`
+        //   that gets the writer used to write to it; should I link these up outside of the csv models?
```

Comment:
Without having looked at the detail, my initial thought is that it likely isn't worth it, but I could be wrong if it happens to be easy and/or useful. @atorralba / @smowton  Thoughts on this?

Reply:
My usual rule of thumb in cases like this is "do it if you can find a real-world use case in which this is relevant", but I know that this can be hard to do sometimes (especially for summaries).

In my opinion, this could actually have more value as an XSS sink rather than as a summary -- intuitively I would say that it's not common to use `HttpServletResponse` as a data container: you are either reading from it (source) or writing to it (sink).

Reply:
A few examples I found in the wild seem to have people logging the result of `getContentAsByteArray`; should we worry about that case at all? There seems to be little point in using a `ContentCachingRequestWrapper` unless the content is used elsewhere.

Reply:
Well, there's [Log Injection/Log Forging](https://owasp.org/www-community/attacks/Log_Injection) -- but what I meant is that `getContentAsByteArray` is probably more of a source (data entering the application from the network) rather than a summary (the application wrote tainted data to the response and later read that same data from it), right?

Nonetheless, if those in-the-wild examples you found show the need of modelling the `response.getWriter().write(tainted)` -> `sink(response.getContentsAsByteArray())` flow, and you think it's easy to do, for me it could make sense.

Reply:
The issue with a step like `response.getWriter().write(tainted) -> sink(response.getContentsAsByteArray())` is that it's going to involve a little bit of ad-hoc backwards flow / aliasing, so it's not something for which we have a fixed solution. Whether to support it at all and how to model such a step very much depends on the use-cases, and it'll never be quite right.

Reply:
Does Java also reuse these types as the response from a client request? "The wild" for me is tabnine---not sure how useful that actually is in determining actual usage. That said, a brief check of the uses of `HttpServletResponse` seems to suggest that it's only used as a response that a server writes to.

I suppose the question is whether the step is worth the hassle when 90% of cases should be caught by the XSS vulnerability that occurs due to tainted data being written to a response in the first place. On the other hand, log injection is still definitely possible with an HTML escaped string.

Most cases I've seen have helper functions that get out the content of a request, and that might not be caught if I did my naive idea of doing local flow the way we do for Go, so this may not be worth doing anyway.

Reply:
It seems we are leaning towards the "not worth it" option, then -- specially if we don't have a supported solution and it would involve a lot of ad-hoc logic. Maybe this is something that should be handled by specific queries, if needed (e.g. trying to model a CVE the flow of which contains this step).